### PR TITLE
Create clang-format config file

### PR DESCRIPTION
**Pull request by djarek**
Due to inconsistencies in code style we have introduce tooling to end
their existence once and for all. I've tried to mimic existing style as closely as possible to minimize required changes. I propose we reformat gradually to reduce the chaos that occurs :D.

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+SpaceBeforeParens: ControlStatements
+SpaceInEmptyParentheses: false
+SpacesBeforeTrailingComments: 1
+SpacesInAngles:  false
+SpacesInContainerLiterals: true
+SpacesInCStyleCastParentheses: false
+SpacesInParentheses: false
+SpacesInSquareBrackets: false
+Standard:        Cpp11
+TabWidth:        2
```
Comment (ranisalt) [inline .clang-format]: What does that mean?

File: .clang-format [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ObjCSpaceBeforeProtocolList: true
+PenaltyBreakBeforeFirstCallParameter: 19
+PenaltyBreakComment: 300
+PenaltyBreakFirstLessLess: 120
+PenaltyBreakString: 1000
+PenaltyExcessCharacter: 1000000
+PenaltyReturnTypeOnItsOwnLine: 60
+PointerAlignment: Left
+ReflowComments:  true
+SortIncludes:    true
```
Comment (ranisalt) [inline .clang-format]: We usually don't do that, mostly we sort as:
- PCH
- STL
- system headers
- project headers

File: .clang-format [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+IndentWidth:     -2
+IndentWrappedFunctionNames: false
+KeepEmptyLinesAtTheStartOfBlocks: true
+MacroBlockBegin: ''
+MacroBlockEnd:   ''
+MaxEmptyLinesToKeep: 1
+NamespaceIndentation: None
+ObjCBlockIndentWidth: 2
+ObjCSpaceAfterProperty: false
+ObjCSpaceBeforeProtocolList: true
```
Comment (ranisalt) [inline .clang-format]: Remove this ObjC configuration as it's not applied.

File: .clang-format [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  AfterStruct:     false
+  AfterUnion:      false
+  BeforeCatch:     false
+  BeforeElse:      false
+  IndentBraces:    false
+BreakBeforeBinaryOperators: None
+BreakBeforeBraces: Custom
+BreakBeforeTernaryOperators: true
+BreakConstructorInitializersBeforeComma: false
+ColumnLimit:     120
```
Comment (ranisalt) [inline .clang-format]: Why? I usually keep it below 80 for two tiled screens side by side.

**Comment by ranisalt** [comment]
It should make the minimal amount of changes to the current source code, did you apply against our current source?

**Comment by ranisalt** [comment]
I also think we should adequate code to the [C++ core guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines)

File: .clang-format (hunk 1 above)
Comment (djarek) [inline .clang-format]: I think all indentation in clang-format is calculated using spaces and if you choose to use tabs you have to specify how many tabs to put per indent. Long story short if this isn't smaller or equal to indent width weird things happen.

File: .clang-format (hunk 2 above)
Comment (djarek) [inline .clang-format]: The ordering of includes can be configured in IncludeCategories. Though you're right, this is likely to break the build in some cases.

File: .clang-format (hunk 4 above)
Comment (djarek) [inline .clang-format]: My experience so far with the 80 character limit is that most people try to work around it by choosing shorter identifier names. Overall I agree that short columns are easier to read but the gain is far outweighed by the fact that people produce stuff like:
```
ret = add_chr_to_srv_inc_c(s, c, TRUE, FALSE, TRUE);
```
I've also noticed that 120 made people worry less about fitting in the line and care more about code readability. My findings may be skewed due to small test group and the fact that people who follow the 80 character rule are quite often very anal and religious about it (to the point where common sense and logic can't touch them).

**Comment by djarek** [comment]
Ok, no [problem](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#a-namerl-camelanl10-avoid-camelcase) :D Introducing more checks from core guidelines(clang-tidy) will have to be done in a separate PR. I did run it agains some shorter source files and it look sort of fine. I'll try to run it on some abomination like protocolgame and see what happens :D

File: .clang-format (hunk 4 above)
Comment (ranisalt) [inline .clang-format]: Well, that's why I'm so pedantic with names on PRs... but I agree.

File: .clang-format (hunk 1 above)
Comment (ranisalt) [inline .clang-format]: Can't it be removed with similar effect?

**Comment by djarek** [comment]
I've checked the config on protocolgame.cpp and had to make some minor adjustments.

**Comment by ranisalt** [comment]
Why don't we base on LLVM?

File: .clang-format [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  AfterFunction:   true
+  AfterNamespace:  false
+  AfterObjCDeclaration: false
+  AfterStruct:     true
+  AfterUnion:      false
+  BeforeCatch:     false
+  BeforeElse:      false
+  IndentBraces:    false
+BreakBeforeBinaryOperators: None
+BreakBeforeBraces: Custom
```
Comment (ranisalt) [inline .clang-format]: We use the Stroustroup, in general.

**Comment by djarek** [comment]
```
# BasedOnStyle:  LLVM
```
The template was generated based on LLVM, but I made changes to reflect the current style. Not sure if this is what you were asking for.

File: .clang-format [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+BreakBeforeTernaryOperators: true
+BreakConstructorInitializersBeforeComma: false
+ColumnLimit:     120
+CommentPragmas:  '^ IWYU pragma:'
+ConstructorInitializerAllOnOneLineOrOnePerLine: false
```
Comment (ranisalt) [inline .clang-format]: I'd like that we set it to true and follow. It definitely adds up readability.

File: .clang-format [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+ConstructorInitializerIndentWidth: 4
+ContinuationIndentWidth: 4
+Cpp11BracedListStyle: true
+DerivePointerAlignment: false
+DisableFormat:   false
+ExperimentalAutoDetectBinPacking: false
+ForEachMacros:   [ foreach, Q_FOREACH, BOOST_FOREACH ]
```
Comment (ranisalt) [inline .clang-format]: This isn't used.

File: .clang-format [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+BreakBeforeTernaryOperators: true
+BreakConstructorInitializersBeforeComma: false
+ColumnLimit:     120
+CommentPragmas:  '^ IWYU pragma:'
+ConstructorInitializerAllOnOneLineOrOnePerLine: false
+ConstructorInitializerIndentWidth: 4
+ContinuationIndentWidth: 4
+Cpp11BracedListStyle: true
+DerivePointerAlignment: false
+DisableFormat:   false
```
Comment (ranisalt) [inline .clang-format]: This isn't used, can be removed if defaults to true.

File: .clang-format [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    Priority:        1
+IndentCaseLabels: true
+IndentWidth:     4
+IndentWrappedFunctionNames: false
+KeepEmptyLinesAtTheStartOfBlocks: true
+MacroBlockBegin: ''
+MacroBlockEnd:   ''
+MaxEmptyLinesToKeep: 1
+NamespaceIndentation: None
+PenaltyBreakBeforeFirstCallParameter: 19
```
Comment (ranisalt) [inline .clang-format]: Magic number?

File: .clang-format [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+PenaltyBreakComment: 300
+PenaltyBreakFirstLessLess: 120
+PenaltyBreakString: 1000
+PenaltyExcessCharacter: 1000000
```
Comment (ranisalt) [inline .clang-format]: Isn't that too harsh for current code base?

File: .clang-format [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+SpaceInEmptyParentheses: false
+SpacesBeforeTrailingComments: 1
+SpacesInAngles:  false
+SpacesInContainerLiterals: true
+SpacesInCStyleCastParentheses: false
+SpacesInParentheses: false
+SpacesInSquareBrackets: false
+Standard:        Cpp11
+TabWidth:        4
+UseTab:          Always
```
Comment (ranisalt) [inline .clang-format]: I prefer `ForIndentation`, for we use spaces for alignment after indentation.

**Comment by ranisalt** [comment]
@djarek but enabling that setting won't base on LLVM and we only need to change what we have different, instead of having the entire format file?

**Comment by djarek** [comment]
@ranisalt 
I'd prefer using one of the predefined styles, but that seems a little extreme :D. I'll see if choosing a base and overriding only some works properly.

**Comment by ranisalt** [comment]
I personally don't fully like any of them :/

File: .clang-format (hunk 5 above)
Comment (djarek) [inline .clang-format]: Stroustrup places else on a separate line:
```
if () {
}
else {
}

File: .clang-format (hunk 9 above)
Comment (djarek) [inline .clang-format]: Looks better with 20. Can't get it less magical :D

File: .clang-format (hunk 10 above)
Comment (djarek) [inline .clang-format]: No mercy to long lines.

**Comment by djarek** [comment]
@ranisalt Anything else?

**Pull request merged**